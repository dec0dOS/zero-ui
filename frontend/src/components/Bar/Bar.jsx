import logo from "./assets/logo.png";

import { useState } from "react";
import { Link as RouterLink, useHistory } from "react-router-dom";
import { useLocalStorage } from "react-use";

import {
  AppBar,
  Toolbar,
  Typography,
  Box,
  Button,
  Divider,
  Menu,
  MenuItem,
  Link,
} from "@material-ui/core";
import MenuIcon from "@material-ui/icons/Menu";

import LogIn from "components/LogIn";

function Bar() {
  const [loggedIn, setLoggedIn] = useLocalStorage("loggedIn", false);
  const [anchorEl, setAnchorEl] = useState(null);

  const history = useHistory();

  const openMenu = (event) => {
    setAnchorEl(event.currentTarget);
  };

  const closeMenu = () => {
    setAnchorEl(null);
  };

  const onLogOutClick = () => {
    setLoggedIn(false);
    localStorage.clear();
    history.push("/");
    history.go(0);
  };

  const onClickGoToManagment = () => {
    history.push("/user-managment");
  };

  const menuItems = [
    {
      name: "User Managment",
      to: "/user-managment",
      onclick: onClickGoToManagment,
    },
    {
      name: "Log out",
      divide: true,
      onClick: onLogOutClick,
    },
  ];

  return (
    <AppBar
      color="secondary"
      style={{ background: "#000000" }}
      position="static"
    >
      <Toolbar>
        <Box display="flex" flexGrow={1}>
          <Typography color="inherit" variant="h6">
            <Link
              color="inherit"
              component={RouterLink}
              to="/"
              underline="none"
            >
              <img src={logo} width="100" height="100" alt="logo" />
            </Link>
          </Typography>
        </Box>

        {loggedIn && (
          <>
            <Button color="inherit" onClick={openMenu}>
              <MenuIcon></MenuIcon>
            </Button>

            <Menu
              anchorEl={anchorEl}
              open={Boolean(anchorEl)}
              onClose={closeMenu}
            >
              {menuItems.map((menuItem, index) => {
                if (
                  menuItem.hasOwnProperty("condition") &&
                  !menuItem.condition
                ) {
                  return null;
                }

                let component = null;

                if (menuItem.to) {
                  component = (
                    <MenuItem
                      key={index}
                      component={RouterLink}
                      to={menuItem.to}
                      onClick={closeMenu}
                    >
                      {menuItem.name}
                    </MenuItem>
                  );
                } else {
                  component = (
                    <MenuItem
                      key={index}
                      onClick={() => {
                        closeMenu();

                        menuItem.onClick();
                      }}
                    >
                      {menuItem.name}
                    </MenuItem>
                  );
                }

                if (menuItem.divide) {
                  return (
                    <span key={index}>
                      <Divider />

                      {component}
                    </span>
                  );
                }

                return component;
              })}
            </Menu>
          </>
        )}
        {!loggedIn && LogIn()}
      </Toolbar>
    </AppBar>
  );
}

export default Bar;
