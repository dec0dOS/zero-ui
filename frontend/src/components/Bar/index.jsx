export { default } from "./Bar.jsx";
