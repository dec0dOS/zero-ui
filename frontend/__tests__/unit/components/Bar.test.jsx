import { render, screen } from "@testing-library/react";
import Bar from "components/Bar";
import { MemoryRouter, Route, Router } from "react-router-dom";
import { createMemoryHistory } from "history";
import userEvent from "@testing-library/user-event";

// Useful reference: https://bholmes.dev/blog/mocking-browser-apis-fetch-localstorage-dates-the-easy-way-with-jest/

let mockStorage = {};

describe("Bar", () => {
  beforeAll(() => {
    global.Storage.prototype.getItem = jest.fn((key) => mockStorage[key]);
    global.Storage.prototype.clear = jest.fn(() => (mockStorage = {}));
  });

  beforeEach(() => {
    mockStorage = {};
  });

  afterAll(() => {
    global.Storage.prototype.getItem.mockReset();
  });

  it("renders Bar unchanged when logged out", () => {
    const history = createMemoryHistory();
    mockStorage["loggedIn"] = false;

    const { container } = render(
      <Router history={history}>
        <Bar />
      </Router>
    );
    expect(container).toMatchSnapshot();
  });

  it("renders Bar unchanged when logged in", () => {
    const history = createMemoryHistory();
    mockStorage["loggedIn"] = true;

    const { container } = render(
      <Router history={history}>
        <Bar />
      </Router>
    );
    expect(container).toMatchSnapshot();
  });

  it("renders Bar unchanged when logged in, after clicking button to open menu", async () => {
    const history = createMemoryHistory();
    const user = userEvent.setup();
    mockStorage["loggedIn"] = true;

    const { container } = render(
      <Router history={history}>
        <Bar />
      </Router>
    );

    const menuButton = screen.getByRole("button", {
      name: "",
    });

    await user.click(menuButton);

    const logOutMenuItem = screen.getByRole("menuitem", {
      name: "Log out",
    });
    expect(logOutMenuItem).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  test("open and close menu when logged in", async () => {
    const history = createMemoryHistory();
    const user = userEvent.setup();
    mockStorage["loggedIn"] = true;

    render(
      <Router history={history}>
        <Bar />
      </Router>
    );

    const menuButton = screen.getByRole("button", {
      name: "",
    });

    // Open menu
    await user.click(menuButton);

    const logOutMenuItem = screen.getByRole("menuitem", {
      name: "Log out",
    });
    expect(logOutMenuItem).toBeInTheDocument();

    // Close menu
    await user.keyboard("{Escape}");

    expect(logOutMenuItem).not.toBeInTheDocument();
  });

  test("open menu and click on 'Log out'", async () => {
    const user = userEvent.setup();
    mockStorage["loggedIn"] = true;
    let testLocation;

    render(
      <MemoryRouter initialEntries={["/test-url/"]}>
        <Route path="/test-url">
          <Bar />
        </Route>
        <Route
          path="*"
          render={({ location }) => {
            testLocation = location;
            return null;
          }}
        />
      </MemoryRouter>
    );

    const menuButton = screen.getByRole("button", {
      name: "",
    });

    // Open menu
    await user.click(menuButton);

    const logOutMenuItem = screen.getByRole("menuitem", {
      name: "Log out",
    });
    expect(logOutMenuItem).toBeInTheDocument();

    // Click on 'Log out'
    await user.click(logOutMenuItem);

    expect(Object.keys(mockStorage).length).toBe(0);
    expect(testLocation.pathname).toBe("/");
  });
});
