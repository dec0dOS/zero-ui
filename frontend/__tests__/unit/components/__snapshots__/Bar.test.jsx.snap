// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bar renders Bar unchanged when logged in 1`] = `
<div>
  <header
    class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorSecondary MuiPaper-elevation4"
    style="background: rgb(0, 0, 0);"
  >
    <div
      class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
    >
      <div
        class="MuiBox-root MuiBox-root-2"
      >
        <h6
          class="MuiTypography-root MuiTypography-h6 MuiTypography-colorInherit"
        >
          <a
            class="MuiTypography-root MuiLink-root MuiLink-underlineNone MuiTypography-colorInherit"
            href="/"
          >
            <img
              alt="logo"
              height="100"
              src="test-file-stub"
              width="100"
            />
          </a>
        </h6>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-colorInherit"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </header>
</div>
`;

exports[`Bar renders Bar unchanged when logged in, after clicking button to open menu 1`] = `
<div
  aria-hidden="true"
>
  <header
    class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorSecondary MuiPaper-elevation4"
    style="background: rgb(0, 0, 0);"
  >
    <div
      class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
    >
      <div
        class="MuiBox-root MuiBox-root-3"
      >
        <h6
          class="MuiTypography-root MuiTypography-h6 MuiTypography-colorInherit"
        >
          <a
            class="MuiTypography-root MuiLink-root MuiLink-underlineNone MuiTypography-colorInherit"
            href="/"
          >
            <img
              alt="logo"
              height="100"
              src="test-file-stub"
              width="100"
            />
          </a>
        </h6>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-colorInherit"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        >
          <span
            class="MuiTouchRipple-ripple MuiTouchRipple-rippleVisible"
            style="width: 2.8284271247461903px; height: 2.8284271247461903px; top: -1.4142135623730951px; left: -1.4142135623730951px;"
          >
            <span
              class="MuiTouchRipple-child MuiTouchRipple-childLeaving"
            />
          </span>
        </span>
      </button>
    </div>
  </header>
</div>
`;

exports[`Bar renders Bar unchanged when logged out 1`] = `
<div>
  <header
    class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorSecondary MuiPaper-elevation4"
    style="background: rgb(0, 0, 0);"
  >
    <div
      class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
    >
      <div
        class="MuiBox-root MuiBox-root-1"
      >
        <h6
          class="MuiTypography-root MuiTypography-h6 MuiTypography-colorInherit"
        >
          <a
            class="MuiTypography-root MuiLink-root MuiLink-underlineNone MuiTypography-colorInherit"
            href="/"
          >
            <img
              alt="logo"
              height="100"
              src="test-file-stub"
              width="100"
            />
          </a>
        </h6>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Log In
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </header>
</div>
`;
